backends/p4tools/testgen/core/small_step/expr_stepper.cpp

```diff
@@ -0,0 +1,475 @@
+#include "backends/p4tools/testgen/core/small_step/expr_stepper.h"
+
+#include <cstddef>
+#include <list>
+#include <utility>
+#include <vector>
+
+#include "backends/p4tools/common/lib/formulae.h"
+#include "backends/p4tools/common/lib/ir.h"
+#include "backends/p4tools/common/lib/saturation_elim.h"
+#include "backends/p4tools/common/lib/symbolic_env.h"
+#include "lib/error.h"
+#include "lib/exceptions.h"
+#include "lib/null.h"
+#include "lib/safe_vector.h"
+
+#include "backends/p4tools/testgen/core/small_step/table_stepper.h"
+#include "backends/p4tools/testgen/lib/continuation.h"
+#include "backends/p4tools/testgen/lib/exceptions.h"
+#include "backends/p4tools/testgen/lib/execution_state.h"
+#include "backends/p4tools/testgen/lib/gen_eq.h"
+
+namespace P4Tools {
+
+namespace P4Testgen {
+
+ExprStepper::ExprStepper(ExecutionState& state, AbstractSolver& solver,
+                         const ProgramInfo& programInfo)
+    : AbstractStepper(state, solver, programInfo) {}
+
+bool ExprStepper::preorder(const IR::BoolLiteral* boolLiteral) {
+    logStep(boolLiteral);
+    return stepSymbolicValue(boolLiteral);
+}
+
+bool ExprStepper::preorder(const IR::Constant* constant) {
+    logStep(constant);
+    return stepSymbolicValue(constant);
+}
+
+void ExprStepper::handleHitMissActionRun(const IR::Member* member) {
+    auto* nextState = new ExecutionState(state);
+    std::vector<Continuation::Command> replacements;
+    const auto* method = member->expr->to<IR::MethodCallExpression>();
+    BUG_CHECK(method->method->is<IR::Member>(), "Method apply has unexpected format: %1%", method);
+    replacements.emplace_back(new IR::MethodCallStatement(method));
+    const auto* methodName = method->method->to<IR::Member>();
+    const auto* table = state.getTableType(methodName);
+    CHECK_NULL(table);
+    if (member->member.name == IR::Type_Table::hit) {
+        replacements.emplace_back(Continuation::Return(TableStepper::getTableHitVar(table)));
+    } else if (member->member.name == IR::Type_Table::miss) {
+        replacements.emplace_back(
+            Continuation::Return(new IR::LNot(TableStepper::getTableHitVar(table))));
+    } else {
+        replacements.emplace_back(Continuation::Return(TableStepper::getTableActionVar(table)));
+    }
+    nextState->replaceTopBody(&replacements);
+    result->emplace_back(nextState);
+}
+
+bool ExprStepper::preorder(const IR::Member* member) {
+    logStep(member);
+
+    if (member->expr->is<IR::MethodCallExpression>() &&
+        (member->member.name == IR::Type_Table::hit ||
```

Comment:
I hope there is no confusion with other methods that can have the same name.
you are not checking that these are applied to a table.

Reply:
Is there another methodCallExpression that can reduce produce .hit? Otherwise we would need to check for the type.

Reply:
it depends on the architectural model and whether it is extensible. In ebpf you can declare your own extern functions (but not methods). So I guess this is good enough for now.